specifyweb/businessrules/migrations/0002_default_unique_rules.py

```diff
@@ -0,0 +1,273 @@
+from django.db import migrations
+from django.core.exceptions import ObjectDoesNotExist, MultipleObjectsReturned
+
+from specifyweb.specify import models as spmodels
+from specifyweb.businessrules.models import UniquenessRule
+
+DEFAULT_UNIQUENESS_RULES = {
```

Comment:
Are the changes in https://github.com/specify/specify7/pull/4215/commits/5ddb6d8bb954f1c6788c57ab93c6952daad3f0d9 sufficient? 

At the time of running the migration, it will read the rules from the flat file and populate the database with those. 
Thus, further edits to the file would have no effect unless the migration is run again.

Reply:
> Thus, further edits to the file would have no effect unless the migration is run again.

how is that a positive thing and not a drawback?

you are needlessly storing things in the database that can be gotten from the .json file directly, and you are preventing us from easily modifying the defaults in the future
unless I'm missing something, the only benefit I see for savings defaults into the database is that fetching rules then is just "fetch from database" rather than "fetch from database, and concat these json rules too" (Though the 2nd doesn't seem much more complicated)

Reply:
also, sorry if this is making it overcomplicated, but can the rules that are isDatabaseConstraint be inferred in some way rather than hard-coded?
if so, that would prevent the need for hardcoding them, thus automate this and reduce possible bugs
if it can be inferred but is expensive, that you could potentially write a test that would infer them, and compare that infered to the stored json - this way the performance cost is paid by the test, not by runtime

Reply:
There might be a few concerns with concatenating the json rules directly with the user defined rules. 

From a user perspective, the primary concern is how we should handle the default rules which are _not_ database constraints. The problem here comes down to flexibility. It is easiest from a developmental standpoint to make all default rules readonly and not allow their further configuration/removal. There is value however, in allowing the rules which are not database constraints to be deleted and modified. 

To alleviate this, I would first imagine the non database uniqueness constraints being included in the initial migration and then subsequently read and concat the database constraints. However, this still does come with the issue wherein further non-readonly uniqueness rules could not be created without a database constraint. 

Another solution is to directly read from the json file and when a non-database-constraint rule is changed, then update the associated rule in the file. This leads to the most complexity, as there is a schism between "types" of rules (default and otherwise). There is an additional concern with this approach in that if the database is sourced to another instance, it will lack the uniqueness rules file and if any default rules were modified those changes would be lost. This would be alleviated by making the rules an SpAppResource, but a solution would be needed about how to handle updates to the file going further, without overriding any edits that have been made to the non-database constraint rules.

Reply:
> can the rules that are isDatabaseConstraint be inferred in some way rather than hard-coded?

This might be possible. 

- We can use SQLAlchemy reflection to fetch constraints: https://docs.sqlalchemy.org/en/20/core/reflection.html#sqlalchemy.engine.reflection.Inspector.get_unique_constraints

- We can directly run a SQL query to fetch the unique columns, as in https://github.com/specify/specify7/issues/2924#issuecomment-1440638103

An approach through Django may be the most difficult option, as the constraints are not added to the Django models when they are created. Perhaps the `inspectdb` management command may create the models with the proper uniqueness constraints? (https://docs.djangoproject.com/en/3.2/howto/legacy-databases/)

Reply:
thanks. it looks like you given good though to this and come to conclusion that the approach you have is the best option.

I also didn't pay enough attention to the fact that there are default read-only rules, default non-read-only rules, and user-added rules (which must not conflict with default read-only rules). Given this, the approach makes more sense.

> This might be possible

Feel free to explore this, but not urgent, as most bugs thus far have been on the front-end, not the back-end (might have to do with the fact that front-end is more than 4 times larger, and is changing rapidly all the time)